### ISSUE TITLE
Signature checking always fails on first install attempt after download

### ISSUE THREAD

**Issue by sdijoseph** [NetSparkleUpdater/NetSparkle#512]
I'm running into an issue where if an install immediately follows a download, the signature checking always fails because the installer file is locked. A repeated install attempt succeeds. Is anyone else running into this issue?

This happens on version "[2.3.0-preview20230704001](https://www.nuget.org/packages/NetSparkleUpdater.SparkleUpdater/2.3.0-preview20230704001)" and "[2.2.3](https://www.nuget.org/packages/NetSparkleUpdater.SparkleUpdater/2.2.3)". It also happens on Windows with an InnoSetup .exe installer and on Linux with a .deb file.

SparkleUpdater code:

```
_sparkle = new SparkleUpdater(
    @"<redacted>", // link to your app cast file
    new Ed25519Checker(SecurityMode.Strict, // security mode -- use .Unsafe to ignore all signature checking (NOT recommended!!)
        "<redacted>") // your base 64 public key -- generate this with the NetSparkleUpdater.Tools.AppCastGenerator .NET CLI tool on any OS
) {
    UIFactory = new NetSparkleUpdater.UI.Avalonia.UIFactory(null), // or null or choose some other UI factory or build your own!
    RelaunchAfterUpdate = true, // default is false; set to true if you want your app to restart after updating (keep as false if your installer will start your app for you)
    // CustomInstallerArguments = "", // set if you want your installer to get some command-line args
    ShowsUIOnMainThread = true, // required on Avalonia, preferred on WPF/WinForms
    AppCastDataDownloader = new LocalFileAppCastDownloader(),
    UpdateDownloader = new LocalFileDownloader()
    {
        LogWriter = new LogWriter(true)
    },
    CustomInstallerArguments = "/SILENT",
    UseNotificationToast = true
};
_sparkle.StartLoop(false, checkFrequency: TimeSpan.FromSeconds(10)); // `true` to run an initial check online -- only call StartLoop once for a given SparkleUpdater instance!
```

Exception in the debug log (redacted installer name and user):

```
netsparkle: File is corrupt or signature is Unchecked; deleting file and redownloading...
netsparkle: Starting to download <installer-name> to C:\Users\<user>\AppData\Local\Temp\<installer-name>
netsparkle: Fully downloaded file exists at C:\Users\<user>\AppData\Local\Temp\<installer-name>
netsparkle: Performing signature check
netsparkle: Error validating signature of file: The process cannot access the file 'C:\Users\<user>\AppData\Local\Temp\<installer-name>' because it is being used by another process.;    at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.SyncWindowsFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategy(FileStream fileStream, String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, Boolean useAsync)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at System.IO.File.OpenRead(String path)
   at NetSparkleUpdater.SignatureVerifiers.Ed25519Checker.VerifySignatureOfFile(String signature, String binaryPath)
   at NetSparkleUpdater.SparkleUpdater.OnDownloadFinished(Object sender, AsyncCompletedEventArgs e)
```

**Comment by Deadpikle** [comment]
Hi @sdijoseph,

Thanks for the bug report. I think this may stem from not closing files properly when they're done downloading:

https://github.com/NetSparkleUpdater/NetSparkle/blob/a38b203a58c97ff43d43d8dd3ef801590b0048b7/src/NetSparkle/Downloaders/WebClientFileDownloader.cs#L164-L198

The progress calls should probably happen after disposal happens.

More reading: https://stackoverflow.com/a/26741192/3938401

It might be a few days before I can get to this with another preview build. You can temporarily get around this in your build by implementing your own copy of `IUpdateDownloader` with a copy of `WebClientFileDownloader` and the appropriate adjustments (if this is indeed the cause of the issue).

Thanks! :-)

**Comment by sdijoseph** [comment]
> Hi @sdijoseph,
> 
> Thanks for the bug report. I think this may stem from not closing files properly when they're done downloading:
> 
> The progress calls should probably happen after disposal happens.
> 
> More reading: https://stackoverflow.com/a/26741192/3938401
> 
> It might be a few days before I can get to this with another preview build. You can temporarily get around this in your build by implementing your own copy of `IUpdateDownloader` with a copy of `WebClientFileDownloader` and the appropriate adjustments (if this is indeed the cause of the issue).
> 
> Thanks! :-)

I actually think I just found the issue. I was using LocalFileDownloader, and the issue is that the DownloadFileCompleted.Invoke is happening inside the using block, so the destinationStream has not been closed at that point. I made my own "FixedLocalFileDownloader" with just that Invoke method moved outside the using block, and it got rid of the error.

https://github.com/NetSparkleUpdater/NetSparkle/blob/a38b203a58c97ff43d43d8dd3ef801590b0048b7/src/NetSparkle/Downloaders/LocalFileDownloader.cs#L107-L132

I think WebClientFileDownloader probably doesn't have this issue actually, as there are explicit Close() calls happening before the DownloadFileCompleted invocation. My suggestion would be to move the DownloadFileCompleted?.Invoke outside the using blocks for all FileDownloaders (and get rid of unneeded Close() calls). 

I can put this in a PR myself if you want, should I merge it into the develop branch?
